Fix names conflict when lifting

RabbitWhite1 opened the pull request:
## Bug description
When pending args that are potentially to be lift [here](https://github.com/pytorch/pytorch/blob/58f346c874a8a982679b4d4f3876602cc05d66d4/torch/_dynamo/output_graph.py#L1866) having same base name, like `contiguous` and `contiguous_1`, the call into [create_graph_input](https://github.com/pytorch/pytorch/blob/58f346c874a8a982679b4d4f3876602cc05d66d4/torch/_dynamo/output_graph.py#L2081) can finally create a name ([here](https://github.com/pytorch/pytorch/blob/58f346c874a8a982679b4d4f3876602cc05d66d4/torch/fx/graph.py#L1008)) that overwrite args to lift. And thus causing a wrong output of graph.

## Reproducing
Below is an reproduceable example, 
```python
import logging
from typing import List

import torch
from functorch.compile import aot_module_simplified, make_boxed_func


@torch.library.custom_op("mylib::somefunc_forward", mutates_args=())
def somefunc_forward(
    input_: torch.Tensor,
    weight: torch.Tensor,
    shape: List[int],
) -> torch.Tensor:
    return torch.ones_like(input_)


@somefunc_forward.register_fake
def _(input_, shape, weight):
    return torch.empty_like(input_)


@torch.library.custom_op("mylib::somefunc_backward", mutates_args=())
def somefunc_backward(
    grad_output: torch.Tensor,
    input_: torch.Tensor,
    weight: torch.Tensor,
    shape: List[int],
) -> torch.Tensor:
    print(f"backward.{grad_output.shape=}")
    print(f"backward.{input_.shape=}")
    print(f"backward.{weight.shape=}")
    print(f"backward.{shape=}")
    assert list(weight.shape) == shape
    return torch.ones_like(weight)


@somefunc_backward.register_fake
def _(grad_output, input_, weight, shape):
    return torch.empty_like(weight)


def a_func(grad_output, input_, weight_, shape):
    return torch.ones_like(input_.sum() * weight_)


class SomeFunc(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input, weight, normalized_shape):
        ctx.normalized_shape = normalized_shape
        input_ = input.contiguous()
        weight_ = weight.contiguous()
        output = somefunc_forward(input_, weight_, ctx.normalized_shape)
        ctx.save_for_backward(input_, weight_)
        return output

    @staticmethod
    def backward(ctx, grad_output):
        input_, weight_ = ctx.saved_tensors
        # grad_weight = a_func(grad_output, input_, weight_, ctx.normalized_shape)
        grad_weight = somefunc_backward(
            grad_output.contiguous(),
            input_,
            weight_,
            ctx.normalized_shape,
        )
        return None, grad_weight, None


class MyModel(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.weight = torch.nn.Parameter(torch.ones(7))

    def forward(self, x):
        return SomeFunc.apply(x, self.weight, [7])


model = MyModel()
torch._logging.set_logs(dynamo=logging.DEBUG, aot=logging.DEBUG, graph_code=True)


def aot_print_backend(gm, sample_inputs):
    # Forward compiler capture
    def fw(gm, sample_inputs):
        print(f"----- fw")
        gm.print_readable()
        return make_boxed_func(gm.forward)

    # Backward compiler capture
    def bw(gm, sample_inputs):
        print(f"----- bw")
        gm.print_readable()
        return make_boxed_func(gm.forward)

    # Call AOTAutograd
    gm_forward = aot_module_simplified(
        gm, sample_inputs, fw_compiler=fw, bw_compiler=bw
    )
    return gm_forward


model = torch.compile(
    model,
    backend=aot_print_backend,
    dynamic=False,
)
out = model(torch.rand((128, 4, 7)))
out.mean().backward()
```

I can see log that showing calling into create_graph_input like
```log
V0629 02:08:46.839914 8200981504 torch/_dynamo/output_graph.py:2042] [0/0] create_graph_input contiguous (none)
V0629 02:08:46.839998 8200981504 torch/_dynamo/output_graph.py:2042] [0/0] create_graph_input contiguous_1 (none)
```

And the backward graph generate will be like
```log
class GraphModule(torch.nn.Module):
    def forward(self, function_ctx, somefunc_forward_default: "f32[128, 4, 7]", contiguous: "f32[128, 4, 7]", contiguous_1: "f32[7]"):
        contiguous_1 = contiguous
        contiguous_2 = contiguous_1
        
        # No stacktrace found for following nodes
        _set_grad_enabled = torch._C._set_grad_enabled(False)
        
         # File: /Users/bytedance/testtorch/test_custom_op_bug.py:61 in backward, code: grad_output.contiguous(),
        contiguous: "f32[128, 4, 7]" = somefunc_forward_default.contiguous();  somefunc_forward_default = None
        
         # File: /opt/tiger/pytorch/torch/_library/custom_ops.py:506 in __call__, code: return self._opoverload(*args, **kwargs)
        somefunc_backward_default: "f32[7]" = torch.ops.mylib.somefunc_backward.default(contiguous, contiguous_1, contiguous_2, [7]);  contiguous = contiguous_1 = contiguous_2 = None
        
        # No stacktrace found for following nodes
        _set_grad_enabled_1 = torch._C._set_grad_enabled(True)
        return (None, somefunc_backward_default)
```

The original code of `somefunc_backward` takes a input list of `grad_output`, `input_`, `weight` and `shape`, where `weight` should be shape of `torch.Size([7])`. However, in the graph, `contiguous1` and `contiguous_2` are assigned with `contiguous`, this leads to assertion failure I added in `somefunc_backward`.

## Environment
```log
Collecting environment information...
PyTorch version: 2.5.0a0+git0b7e8df
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.5 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: version 3.26.4
Libc version: N/A

Python version: 3.9.19 (main, May  6 2024, 14:39:30)  [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-14.5-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Pro

Versions of relevant libraries:
[pip3] numpy==2.0.0
[pip3] optree==0.11.0
[pip3] torch==2.5.0a0+git0b7e8df
[pip3] torchgraph==0.0.1
[conda] numpy                     2.0.0                    pypi_0    pypi
[conda] optree                    0.11.0                   pypi_0    pypi
[conda] torch                     2.5.0a0+git0b7e8df           dev_0    <develop>
[conda] torchgraph                0.0.1                     dev_0    <develop>
```

## How to fix?

I put a naive fix that add the potential args to lift into the used_names. This visits private variables, will fix that if this issue makes sense to you.

@zou3519 @oulgen 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec @peterbell10

zou3519 commented:
I need to think a little more about this. cc'ed some folks who have worked on this part of the code for more thoughts

zou3519 commented on torch/_dynamo/output_graph.py:
```diff
@@ -1863,6 +1863,9 @@ def create_proxy(
         if self.parent is not None:
             flat_args, tree_spec = pytree.tree_flatten((args, kwargs))
             new_flat_args = []
+            for arg in flat_args:
+                if isinstance(arg, torch.fx.proxy.Proxy):
+                    self.graph._graph_namespace._used_names.add(arg.node.name)
```
Er, is the problem just that we get a name collision?
```
        contiguous_1 = contiguous
        contiguous_2 = contiguous_1
```
The problem sounds like FX's naming logic didn't realize that `contiguous_1` was already used. Would it be possible to solve the problem in torch.fx instead of here (which is for Dynamo)?

RabbitWhite1 replied:
> Er, is the problem just that we get a name collision?

Yes, exactly.

> Would it be possible to solve the problem in torch.fx instead of here (which is for Dynamo)?

Let me try to explain what happens here to better answer your question.

In the example above, the lifting happens when processing this function call
```python
# Let's call this node2. Note that it is not yet generated here.
 somefunc_backward_default: "f32[7]" = torch.ops.mylib.somefunc_backward.default(contiguous, contiguous_1, contiguous_2, [7]);
```
And the graph at this moment looks like
```python
graph():
    %function_ctx : [num_users=0] = placeholder[target=function_ctx]
    %somefunc_forward_default : [num_users=1] = placeholder[target=somefunc_forward_default]
    %_set_grad_enabled : [num_users=0] = call_function[target=torch._C._set_grad_enabled](args = (False,), kwargs = {})
    # Let's call this node1
    %contiguous : [num_users=0] = call_method[target=contiguous](args = (%somefunc_forward_default,), kwargs = {})
```
For the first three args of `node2`, the names of their proxies are `contiguous`, `contiguous` and `contiguous_1` respectively, where the last two directly come from the saved ctx in forward pass.

Thus, when trying to lift the second (`contiguous`), FX has no idea about the existence of `contiguous_1`. And thus Dyanmo creates a placeholder node, where the target is the `contiguous` proxy from forward.
```python
# Let's call it node3
%contiguous_1 : [num_users=0] = placeholder[target=contiguous]
```
Similarly, Dynamo creates a placeholder node for the third arg:
```python
# Let's call it node4
%contiguous_2 : [num_users=0] = placeholder[target=contiguous_1]
```

Thus, `node4` actually uses value from `node3` (`contiguous_1` and  `contiguous_2` actually become same variable) instead of the `contiguous_1` that was lifted to the input args.

As a conclusion, I feel it is hard to do this in FX side because it just doesn't know the args list. But I am new to Dynamo, I may miss something. Feel free to correct me.

zou3519 replied:
Let me try to summarize the problem:
- when we lift a freevar, we use the name of the freevar to create a graph input
- we create the graph input by creating a new FX node with kind="placeholder" and target=name
- it is possible that the target conflicts with some node that already exists in the FX graph.

If that sounds right, I'm not sure the fix in this PR is sufficient. Can you instead put a check in https://github.com/pytorch/pytorch/blob/f9bb258892cc71b863d8eb3686a361150be07ee2/torch/_dynamo/output_graph.py#L2102 to see if the name of the new graph input is already taken, and if it is, then we generate a new unique name?

RabbitWhite1 replied:
For the second point, besides kind and target, the proxy name itself will be generated by FX in
https://github.com/pytorch/pytorch/blob/f9bb258892cc71b863d8eb3686a361150be07ee2/torch/fx/graph.py#L1008
So for the third point, it is not the target conflicting with some nodes that already exists in the FX graph (the target just exists in the FX graph for forward pass, and we are currently generating for backward pass). It is more like the generated node name conflicting with some targets from previous tracer. 

In our case for `node3`, target="contiguous" and `node3`'s name is "contiguous_1" (generated by FX, because the first non-conflicting name for format "contiguous_{num}" is "contiguous_1"). And then when generating `node4`, the target of `node4` is "contiguous_1", which conflicts with `node3`'s name.


Btw, the "previous tracer" in our case is the the `fwd_tracer` here:
https://github.com/pytorch/pytorch/blob/f9bb258892cc71b863d8eb3686a361150be07ee2/torch/_dynamo/variables/higher_order_ops.py#L1656

> Can you instead put a check in...

Thus, putting a check here cannot catch the issue, because "contiguous_1" is not in the FX graph when generating `node3`.
Note that given "contiguous_1" is in the previous FX graph, we may also enforce the FX graph for [bwd_tracer](https://github.com/pytorch/pytorch/blob/f9bb258892cc71b863d8eb3686a361150be07ee2/torch/_dynamo/variables/higher_order_ops.py#L1695) to consider the names of `fwd_tracer`. However, this still need to directly access private variable of FX graph.

zou3519 replied:
> Thus, putting a check here cannot catch the issue, because "contiguous_1" is not in the FX graph when generating node3.

Will a check here catch the issue when generating node4? When generating node4, we can pick a different target (so, instead of "contiguous_1" we could have picked "contiguous_1_1")

RabbitWhite1 replied:
Ah I see what you mean. Yes, this can catch the issue when generating node4. I pushed the fix.

RabbitWhite1 replied:
Instead of original manually finding unique input name, I replace that part of code with graph_namespace.create_name. PTAL.

zou3519 commented on torch/_dynamo/output_graph.py:
```diff
@@ -2061,10 +2065,13 @@ def create_graph_input(self, name, type_expr=None, before=False, source=None):
 
         # unique
-        if name in self.input_name_to_proxy:
+        if name in self.input_name_to_proxy or name in self.illegal_input_names:
```
If we just add `name = self.graph._graph_namespace.create_name(name, None)` somewhere here, then we don't need `self.illegal_input_names`. In general we shouldn't be creating a graph input with a name that already exists

RabbitWhite1 replied:
Thanks. And I think this will be similar to my original fix, which add the names of pending-lifting freevars 
(https://github.com/pytorch/pytorch/pull/129817/commits/b0ae23eb87c4fc3d163e8f836eed9025d06211a6). If this sounds better to you, I can change to `_graph_namespace.create_name`. :)

zou3519 replied:
This seems better to me

RabbitWhite1 replied:
Instead of original manually finding unique input name, I replace that part of code with graph_namespace.create_name. PTAL.

zou3519 commented on torch/_dynamo/output_graph.py:
```diff
@@ -2064,16 +2060,8 @@ def create_graph_input(self, name, type_expr=None, before=False, source=None):
-                    name = candidate_name
-                    break
+        # Use graph_namespace to create unique name.
+        name = self.graph._graph_namespace.create_name(name, obj=None)
```
Can you also add a test? the example in your PR body would go well into test/dynamo/test_autograd_function.py

RabbitWhite1 replied:
Sure!

RabbitWhite1 replied:
Pushed the test. Had to update `test_user_defined_object_as_input` to fit the new naming.

zou3519 reviewed (commented):
looks good, just needs a test

zou3519 commented on test/dynamo/test_autograd_function.py:
```diff
@@ -223,6 +225,69 @@ def forward(self, x):
         return self.f(x)
 
 
+@torch.library.custom_op("mylib::custom_op_forward", mutates_args=())
```
nit: use namespace "_torch_testing" so it doesn't conflict

zou3519 commented:
Btw, you might want to rebase on top of https://github.com/pytorch/pytorch/pull/130271. That one changes how some of the graphs are printed

zou3519 commented:
I see some merge conflicts

RabbitWhite1 commented:
> I see some merge conflicts

fixed

zou3519 commented:
@pytorchbot merge

clee2000 commented:
@pytorchbot revert -m "broke dynamo/test_inline_inbuilt_nn_modules.py https://github.com/pytorch/pytorch/actions/runs/9940532858/job/27461141919 https://hud.pytorch.org/pytorch/pytorch/commit/74da2a467f166e00316aee82ba24835ca563ed87.  Test passed on PR, possibly a landrace?" -c landrace

clee2000 commented:
@pytorchbot merge -f "re merging after false revert sorry"

clee2000 commented:
@pytorchbot revert -m "Failing inductor/test_flex_attention.py https://github.com/pytorch/pytorch/actions/runs/9940532858/job/27478084137 https://hud.pytorch.org/pytorch/pytorch/commit/74da2a467f166e00316aee82ba24835ca563ed87 Sorry for the churn, possibly a landrace?" -c landrace

RabbitWhite1 commented:
Thanks. Looks like there are some more tests need to be fix. Will fix them soon.

RabbitWhite1 commented:
Fixed the [test_flex_attention](https://github.com/pytorch/pytorch/pull/129817/commits/ada23fd32089db4da3e92942965f8e4e5d30c68b). I am not sure if I fixed all the similar naming issues because tests run endlessly on my cluster... May need to use Github actions to run tests again. Sorry for that...

RabbitWhite1 committed:
Format

RabbitWhite1 committed:
Use _graph_namespace.create_name and add test

RabbitWhite1 committed:
Fix confliting names when lift

RabbitWhite1 committed:
Rename custom op namespace

RabbitWhite1 committed:
Check when generating input placeholder target instead

RabbitWhite1 committed:
Rebase to #130271

Fix functorch tests

Fix tests

Fix test_flex_attention

Rebase, noticed graph name changing, doing other testing

rebase to before readable module name

zou3519 reviewed (commented):
So, now that we're here, one question I have is: why does this result in so many name changes? I would expect the change only changes the ambiguous names

RabbitWhite1 commented:
> So, now that we're here, one question I have is: why does this result in so many name changes? I would expect the change only changes the ambiguous names

Ah that's because in the [create_graph_input](https://github.com/RabbitWhite1/pytorch/blob/e460d47b438112cdfde4d437db28e05a728ede73/torch/_dynamo/output_graph.py#L2043) method, if we use `name = self.graph._graph_namespace.create_name(name, obj=None)` to create an unique name for inputs. 
And then when creating the placeholder proxy [here](https://github.com/RabbitWhite1/pytorch/blob/zhanghan/fix-lift/torch/_dynamo/output_graph.py#L2081), it will create a new target name so that it won't conflict with the input name (they use the same graph_namespace).

The fundamental difference here is that the `unique` is only for input names for the original codes, and for all names in the graph for the fixed codes. Go back to the example I proposed in this PR, the order of name creating in the original version is 
```python
# %contiguous_1 : [num_users=0] = placeholder[target=contiguous]
# %contiguous_2 : [num_users=0] = placeholder[target=contiguous_1]
1. target(input name) contiguous
2. node name contiguous_1
3. target(input name) contiguous_1
4. node name contiguous_2
```
For the fixed version is
```
# %contiguous_2 : [num_users=0] = placeholder[target=contiguous_1]
# %contiguous_4 : [num_users=0] = placeholder[target=contiguous_3]
1. target(input name) contiguous_1
2. node name contiguous_2
3. target(input name) contiguous_3
4. node name contiguous_4
```
But what we may want (to make less modification to those tests) is:
```
# %contiguous_3 : [num_users=0] = placeholder[target=contiguous_1]
# %contiguous_2 : [num_users=0] = placeholder[target=contiguous_2]
1. target(input name) contiguous_1
2. node name contiguous_3
3. target(input name) contiguous_2
4. node name contiguous_2
```
Note that this can avoid giving a new name to the input `contiguous_2`. If we want to do in this manner, we may have to first confirm which inputs will be lift and then create the the node (instead of check and create one by one).
